llvm/lib/Target/Xtensa/XtensaISelLowering.cpp

```diff
@@ -713,6 +746,158 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
+  SDValue ShamtMinusRegisterSize =
+      DAG.getNode(ISD::ADD, DL, VT, Shamt, MinusRegisterSize);
+
+  SDValue LoTrue = DAG.getNode(ISD::SHL, DL, VT, Lo, Shamt);
+  SDValue HiTrue = DAG.getNode(XtensaISD::SRCL, DL, VT, Hi, Lo, Shamt);
+  SDValue Zero = DAG.getConstant(0, DL, VT);
+  SDValue HiFalse = DAG.getNode(ISD::SHL, DL, VT, Lo, ShamtMinusRegisterSize);
+
+  SDValue Cond = DAG.getSetCC(DL, VT, ShamtMinusRegisterSize, Zero, ISD::SETLT);
+  Lo = DAG.getNode(ISD::SELECT, DL, VT, Cond, LoTrue, Zero);
+  Hi = DAG.getNode(ISD::SELECT, DL, VT, Cond, HiTrue, HiFalse);
+
+  return DAG.getMergeValues({Lo, Hi}, DL);
+}
+
+SDValue XtensaTargetLowering::LowerShiftRightParts(SDValue Op,
+                                                   SelectionDAG &DAG,
+                                                   bool IsSRA) const {
+  SDLoc DL(Op);
+  SDValue Lo = Op.getOperand(0), Hi = Op.getOperand(1);
+  SDValue Shamt = Op.getOperand(2);
+  MVT VT = MVT::i32;
+
+  // SRA expansion:
+  //   if Shamt - register size < 0: // Shamt < register size
+  //     Lo = (Lo >>u Shamt) | (Hi << u (register size - Shamt))
+  //     Hi = Hi >>s Shamt
+  //   else:
+  //     Lo = Hi >>s (Shamt - register size);
+  //     Hi = Hi >>s (register size - 1)
+  //
+  // SRL expansion:
+  //   if Shamt - register size < 0: // Shamt < register size
+  //     Lo = (Lo >>u Shamt) | (Hi << u (register size - Shamt))
+  //     Hi = Hi >>u Shamt
+  //   else:
+  //     Lo = Hi >>u (Shamt - register size);
+  //     Hi = 0;
+
+  unsigned ShiftRightOp = IsSRA ? ISD::SRA : ISD::SRL;
+  SDValue MinusRegisterSize = DAG.getConstant(-32, DL, VT);
+  SDValue RegisterSizeMinus1 = DAG.getConstant(32 - 1, DL, VT);
+  SDValue ShamtMinusRegisterSize =
+      DAG.getNode(ISD::ADD, DL, VT, Shamt, MinusRegisterSize);
+
+  SDValue LoTrue = DAG.getNode(XtensaISD::SRCR, DL, VT, Hi, Lo, Shamt);
+  SDValue HiTrue = DAG.getNode(ShiftRightOp, DL, VT, Hi, Shamt);
+  SDValue Zero = DAG.getConstant(0, DL, VT);
+  SDValue LoFalse =
+      DAG.getNode(ShiftRightOp, DL, VT, Hi, ShamtMinusRegisterSize);
+  SDValue HiFalse;
+
+  if (IsSRA) {
+    HiFalse = DAG.getNode(ShiftRightOp, DL, VT, Hi, RegisterSizeMinus1);
+  } else {
+    HiFalse = Zero;
+  }
+
+  SDValue Cond = DAG.getSetCC(DL, VT, ShamtMinusRegisterSize, Zero, ISD::SETLT);
+  Lo = DAG.getNode(ISD::SELECT, DL, VT, Cond, LoTrue, LoFalse);
+  Hi = DAG.getNode(ISD::SELECT, DL, VT, Cond, HiTrue, HiFalse);
+
+  return DAG.getMergeValues({Lo, Hi}, DL);
+}
+
+SDValue XtensaTargetLowering::LowerCTPOP(SDValue Op, SelectionDAG &DAG) const {
+  EVT VT = Op->getValueType(0);
+  SDValue Val = Op.getOperand(0);
+  SDLoc DL(Op);
+
+  if (VT != MVT::i32)
+    return SDValue();
+
+  // CTPOP expansion:
+  // Val = (Val - (Val >> 1)) & 0x55555555
+  // Val = ((Val >> 2) & 0x33333333) + (Val & 0x33333333)
+  // Val = ((Val >> 4) + Val) & 0x0f0f0f0f
+  // Val = (Val >> 8) + Val
+  // Val = (extract bits [16, 20] from Val) + Val
+  // Val = extract bits [0, 5] from Val
+
```

Comment:
This looks pretty close to the generic expansion. Can you just pattern match the EXTUIs at the end from the end sequence of the generic expansion in a DAG combine?

Reply:
The issue is that if you set CTPOP action to Expand, CTTZ_ZERO_UNDEF expands via table lookup and we get back to GlobalVariable and ConstantPool mess.

I'd just custom lower CTPOP to a libcall to avoid this code duplication (assuming compiling the Linux kernel is not in the near future).

Reply:
> The issue is that if you set CTPOP action to Expand, CTTZ_ZERO_UNDEF expands via table lookup and we get back to GlobalVariable and ConstantPool mess.
> 
> I'd just custom lower CTPOP to a libcall to avoid this code duplication (assuming compiling the Linux kernel is not in the near future).

You suggestion is to implement call to "__popcountsi2"?

Reply:
It depends on whether you want to use a libcall or not. 

> The issue is that if you set CTPOP action to Expand, CTTZ_ZERO_UNDEF expands via table lookup 

The default expansion logic doesn't always do the right thing but is usually fixable?

Reply:
Custom lowering can also just directly call TLI::expandCTPOP

Reply:
> Custom lowering can also just directly call TLI::expandCTPOP

I implemented call  to "TLI::expandCTPOP" in lowerCTPOP and improved SRL selection (generate EXTUI for suitable constants) and now CTPOP generic expansion generates the same amount of operation as in my previous implementation.